src/pages/p_oauth.php

```diff
@@ -242,13 +243,31 @@ private function instance_response($authi, $tok, $tokdata) {
         if (isset($jid->orcid) && is_string($jid->orcid)) {
             $reg["orcid"] = $jid->orcid;
         }
+        if (isset($jid->affiliation) && is_string($jid->affiliation)) {
+            $reg["affiliation"] = $jid->affiliation;
+        }
         $info = LoginHelper::check_external_login(Contact::make_keyed($this->conf, $reg));
         if (!$info["ok"]) {
             LoginHelper::login_error($this->conf, $jid->email, $info, null);
             throw new Redirection($tokdata->site_uri);
         }
 
         $user = $info["user"];
+        error_log(var_export($authi, true)."\n");
+        if (isset($jid->groups) && isset($authi->group_mappings)) {
+            foreach ($authi->group_mappings as $group => $role) {
+                if (in_array($group, $jid->groups, true)) {
+                    error_log("setting: ".$group." ".constant("Contact::".$role)."\n");
+                    $user_roles = $user->roles | constant("Contact::".$role);
```

Comment:
I love this idea, it's very nice, but using `constant` is kind of gross & might throw an exception on misconfiguration. It would be better to use `UserStatus::$role_map`. I am have made `UserStatus::parse_roles()` usable in this context (in commit 9c978fb3a9607cabc98bed491368d4cd4b4aaae4), so you could call `$user_roles = UserStatus::parse_roles($role, $user->roles)`.

Reply:
Wasn't aware of that. Let me take a shot.